base/ca/src/main/java/com/netscape/ca/CertificateAuthority.java

```diff
@@ -1371,34 +1371,35 @@ public OCSPResponse validate(OCSPRequest request)
          *    CertID in the request.
          *
          * 2. If this CA is *not* the issuer, look up the issuer
-         *    by its DN in CAEngine.  If not found, fail.  If
-         *    found, dispatch to its 'validate' method.  Otherwise
-         *    continue.
+         *    by its DN in CAEngine. If found, dispatch to its 'validate'
+         *    method. Otherwise continue.
          *
          * 3. If this CA is NOT the issuing CA, we locate the
          *    issuing CA and dispatch to its 'validate' method.
          *    Otherwise, we move forward to generate and sign the
          *    aggregate OCSP response.
          */
-        CertificateAuthority ocspCA = this;
-        if (engine.getCAs().size() > 0 && tbsReq.getRequestCount() > 0) {
+        for (CertificateAuthority ocspCA: engine.getCAs()) {
             Request req = tbsReq.getRequestAt(0);
-            BigInteger serialNo = req.getCertID().getSerialNumber();
-
-            CertificateRepository certificateRepository = engine.getCertificateRepository();
-            X509CertImpl cert = certificateRepository.getX509Certificate(serialNo);
-
-            X500Name certIssuerDN = cert.getIssuerName();
-            ocspCA = engine.getCA(certIssuerDN);
-        }
-
-        if (ocspCA == null) {
-            logger.error("CertificateAuthority: Could not locate issuing CA");
-            throw new CANotFoundException("Could not locate issuing CA");
+            CertID cid = req.getCertID();
+            byte[] nameHash = null;
+            String digestName = cid.getDigestName();
+            if (digestName != null) {
+                try {
+                    MessageDigest md = MessageDigest.getInstance(digestName);
+                    nameHash = md.digest(ocspCA.getSubjectObj().getX500Name().getEncoded());
+                } catch (NoSuchAlgorithmException | IOException e) {
+                    logger.info("CertificateAuthority: OCSP request hash algorithm " + digestName + " not recognised - ");
+                }
```

Comment:
This is an unexpected exception, so I'd suggest to log the error message and the stack trace for troubleshooting, e.g.:
```
logger.error("CertificateAuthority: Unable to generate message digest: " + e.getMessage(), e);
```

Suppose we have a request with an invalid `digestName`, what would be the eventual OCSP response? I just want to make sure OCSP won't return a valid response (i.e. good, revoked, unknown)  that can be misinterpreted by the OCSP client.

If there's such a risk, it might be better to just let the exception bubble up (i.e. rethrow the exception) and let the OCSP client handle the failure. But if it's OK to ignore, then maybe we should just generate a warning without the stack trace:

```
logger.warn("CertificateAuthority: Unable to generate message digest: " + e.getMessage());
```

Reply:
The current behaviour is to return "unknown" if the digest does not work. This is done in `processRequest()` method of the same class. This is acceptable for the protocol because the client should handle this case as this responder cannot validate the certificate and other responders should be invoked.
The protocol provide additional error states which were not used. As an improvement we could implement them to be more accurate in the answer but we can do in a separate PR if it is needed.